dev/tools/localization/gen_l10n.dart

```diff
@@ -0,0 +1,318 @@
+    if (match.groupCount == 2) {
+      String argValue = match.group(2);
+      for (String placeholder in placeholders)
+        argValue = argValue.replaceAll('#$placeholder#', '\$$placeholder');
+
+      methodArgs.add("${pluralIds[pluralKey]}: '$argValue'");
+    }
+  }
+
+  return pluralMethodTemplate
+    .replaceAll('@methodName', key)
+    .replaceAll('@methodParameters', genMethodParameters(bundle, key, 'int').join(', '))
+    .replaceAll('@intlMethodArgs', methodArgs.join(',\n      '));
+}
+
+String genSupportedLocaleProperty(List<LocaleInfo> supportedLocales) {
+  const String prefix = 'static const List<Locale> supportedLocales = <Locale>[ \n    Locale(''';
+  const String suffix = '),\n  ];';
+
+  String resultingProperty = prefix;
+
+  for (int index = 0; index < supportedLocales.length; index += 1) {
+    final String languageCode = supportedLocales[index].languageCode;
+    final String countryCode = supportedLocales[index].countryCode;
+    resultingProperty += '\'$languageCode\'';
+
+    if (countryCode != null)
+      resultingProperty += ', \'$countryCode\'';
+
+    if (index < supportedLocales.length - 1)
+      resultingProperty += '),\n    Locale(';
+  }
+  resultingProperty += suffix;
+
+  return resultingProperty;
+}
+
+Future<void> main(List<String> args) async {
+  final argslib.ArgParser parser = argslib.ArgParser();
+  parser.addOption('dir-path', defaultsTo: path.join('lib', 'l10n'));
+  parser.addOption('input-arb-file', defaultsTo: 'app_en.arb');
+  parser.addOption('output-file-prefix', defaultsTo: 'app');
+  parser.addOption('output-class-prefix', defaultsTo: 'App');
+  final argslib.ArgResults results = parser.parse(args);
+
+  final Directory l10nDirectory = Directory(results['dir-path']);
+  final File inputArbFile = File(path.join(l10nDirectory.path, results['input-arb-file']));
+  final File outputFile = File(path.join(l10nDirectory.path, '${results['output-file-prefix']}_localizations.dart'));
+  final String stringsClassName = '${results['output-class-prefix']}Localizations';
+
+  final List<String> arbFilenames = <String>[];
+  final Set<String> supportedLanguageCodes = <String>{};
+  final List<LocaleInfo> supportedLocales = <LocaleInfo>[];
+
+  for (FileSystemEntity entity in l10nDirectory.listSync()) {
+    final String entityPath = entity.path;
+
+    if (FileSystemEntity.isFileSync(entityPath)) {
+      // TODO: what if there are multiple files with the exact same locale?
+      final RegExp arbFilenameRE = RegExp(r'(\w+)\.arb$');
+      if (arbFilenameRE.hasMatch(entityPath)) {
+        final File arbFile = File(entityPath);
+        final Map<String, dynamic> arbContents = json.decode(arbFile.readAsStringSync());
+        final RegExp arbFilenameLocaleRE = RegExp(r'^[^_]*_(\w+)\.arb$');
+        String localeString = arbContents['@@locale'];
+
+        if (arbFilenameLocaleRE.hasMatch(entityPath) && localeString == null)
+          localeString = arbFilenameLocaleRE.firstMatch(entityPath)[1];
+
```

Comment:
If localeString is null at this point then we found an arb file but couldn't determine what locale it was for. That seems bad; probably abort here.